src/tasks/python/PythonExtensionHelper.ts

```diff
@@ -0,0 +1,116 @@
+    }
+
+    export function getDebuggerLogFilePath(folderName: string) : string {
+    // The debugger generates the log file with the name in this format: ptvsd-{pid}.log,
+    // So given that we run the debugger as the entry point, then the PID is guaranteed to be 1.
+        return path.join(os.tmpdir(), folderName, "ptvsd-1.log");
+    }
+
+    export async function ensureDebuggerReady(prelaunchTask: vscode.Task, debuggerSemaphorePath: string, containerName: string, cliDockerClient: CliDockerClient): Promise<void> {
+        // tslint:disable-next-line:promise-must-complete
+        return new Promise((resolve, reject) => {
+            vscode.tasks.onDidEndTask(async e => {
+                if (e.execution.task === prelaunchTask) {
+                    // There is no way to know the result of the completed task, so a best guess is to check if the container is running.
+                    const containerRunning = await cliDockerClient.inspectObject(containerName, { format: "{{.State.Running}}" });
+
+                    if (containerRunning === "false") {
+                        reject("Failed to attach the debugger, please see the terminal output for more details.");
+                    }
+
+                    const maxRetriesCount = 20;
+                    let retries = 0;
+                    let created = false;
+
+                    // Look for the magic string below in the log file with a retry every 0.5 second for a maximum of 10 seconds.
+                    // TODO: Should be gone as soon as the retry logic is part of the Python debugger/extension.
+                    while (++retries < maxRetriesCount && !created) {
+                        if (fse.existsSync(debuggerSemaphorePath)) {
+                            const contents = fse.readFileSync(debuggerSemaphorePath);
+
+                            created = contents.toString().indexOf("Starting server daemon on") >= 0;
+                            if (created) {
+                                break;
+                            }
+                        }
+
+                        await delay(500);
+                    }
+
+                    if (created) {
+                        resolve();
+                    } else {
+                        reject("Failed to attach the debugger within the alotted timeout.");
+                    }
+                }});
+        })
+    }
+
+    export function getRemotePtvsdCommand(target: PythonFileTarget | PythonModuleTarget, args?: string[], options?: DebugLaunchOptions): string {
+        let fullTarget: string;
+
+        if ((target as PythonFileTarget).file) {
+            fullTarget = (target as PythonFileTarget).file;
+        } else if ((target as PythonModuleTarget).module) {
+            fullTarget = `-m ${(target as PythonModuleTarget).module}`;
+        } else {
+            throw new Error("One of either module or file must be provided.");
+        }
+
+        options = options || {};
```

Comment:
Note: we can now use `??` instead of `||` in many (most?) circumstances.  (Though note the implications around, for example, the empty string `''` being used as to indicate the default value.

Reply:
Isn't it `?.` most of the time?

Reply:
Well, yes, we can use `?.` now too, and they can often be used together (e.g. `obj?.prop ?? defaultValue`).

Reply:
It makes sense to use `??` in the above case where we have an object, but wouldn't `||` still be better for strings? Example:
`buildOptions.context = buildOptions.context || '${workspaceFolder}';`

Reply:
Depends on whether you want `''` to indicate "use the default value" or "literally use an empty string".  If the former, then use `||`.  If the latter, use `??`.  The former has traditionally been used as it was just a consequence of using the `||` form, which was too convenient *not* to use and the empty string edge case was just glossed over given that in both JS and JSON, 'undefined' is far more likely to be used for unspecified strings rather than the empty string.

Reply:
Cool, for strings I left it as "||" and for objects I changed it to "??".